src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -152,6 +154,87 @@ private MethodBase FunctionLookup(IList args)
             return mi;
         }
 
+        private static Dictionary<Type, int> GetTypeStatisticsForArray(ExprListNode array)
+        {
+            var usageFreq = new Dictionary<Type, int>();
+
+            foreach (var exp in array.Exprs)
+            {
+                if (exp is ExprListNode eln)
+                {
+                    var subArray = GetTypeStatisticsForArray(eln);
+                    var t = GetOverallTypeForArray(subArray);
+
+                    if (t != typeof(object))
+                    {
+                        t = t.MakeArrayType();
+                    }
+                    if (!usageFreq.ContainsKey(t))
+                    {
+                        usageFreq.Add(t, 0);
+                    }
+                    usageFreq[t] += 1;
+                }
+                else
+                {
+                    Type t;
+                    switch (exp.Kind)
+                    {
+                        case AstKind.Integer:
+                            t = typeof(long);
+                            break;
+                        case AstKind.Double:
+                            t = typeof(double);
+                            break;
+                        case AstKind.Boolean:
+                            t = typeof(bool);
+                            break;
+                        case AstKind.Char:
+                            t = typeof(char);
+                            break;
+                        case AstKind.String:
+                            t = typeof(string);
+                            break;
+                        case AstKind.Identifier:
+                            t = typeof(object);
+                            break;
+                        default:
+                            t = typeof(object);
+                            break;
+                    }
+                    if (!usageFreq.ContainsKey(t))
+                    {
+                        usageFreq.Add(t, 0);
+                    }
+
+                    usageFreq[t] += 1;
+                }
+            }
+            return usageFreq;
+        }
+
+        private static Type GetOverallTypeForArray(Dictionary<Type, int> arrayTypes)
+        {
+            var keys = arrayTypes.Keys;
+            if (keys.Count == 1)
+            {
+                // There is only a single type in the array, return it.
+                return keys.FirstOrDefault();
+            }
+            if(keys.Count == 2)
+            {
+                if(keys.Any(x => x == typeof(long)) && keys.Any(x => x == typeof(double)))
```

Comment:
all of these `.Any` queries will actually iterate through the `keys` array.
Do you think it may be faster to iterate once and check for all these expected types ?
i,e 
bool isLong, isDouble, isChar, isString = false
foreach(Type key in keys) {
   isLong |= key == typeof(long)
   ....
}
Not sure this is going to matter much...